kmp-library/src/commonMain/kotlin/uk/co/lucasweb/aws/v4/signer/hash/Sha256.kt

```diff
@@ -0,0 +1,206 @@
+            -0x164a245b,
+            0x3956c25b,
+            0x59f111f1,
+            -0x6dc07d5c,
+            -0x54e3a12b,
+            -0x27f85568,
+            0x12835b01,
+            0x243185be,
+            0x550c7dc3,
+            0x72be5d74,
+            -0x7f214e02,
+            -0x6423f959,
+            -0x3e640e8c,
+            -0x1b64963f,
+            -0x1041b87a,
+            0x0fc19dc6,
+            0x240ca1cc,
+            0x2de92c6f,
+            0x4a7484aa,
+            0x5cb0a9dc,
+            0x76f988da,
+            -0x67c1aeae,
+            -0x57ce3993,
+            -0x4ffcd838,
+            -0x40a68039,
+            -0x391ff40d,
+            -0x2a586eb9,
+            0x06ca6351,
+            0x14292967,
+            0x27b70a85,
+            0x2e1b2138,
+            0x4d2c6dfc,
+            0x53380d13,
+            0x650a7354,
+            0x766a0abb,
+            -0x7e3d36d2,
+            -0x6d8dd37b,
+            -0x5d40175f,
+            -0x57e599b5,
+            -0x3db47490,
+            -0x3893ae5d,
+            -0x2e6d17e7,
+            -0x2966f9dc,
+            -0xbf1ca7b,
+            0x106aa070,
+            0x19a4c116,
+            0x1e376c08,
+            0x2748774c,
+            0x34b0bcb5,
+            0x391c0cb3,
+            0x4ed8aa4a,
+            0x5b9cca4f,
+            0x682e6ff3,
+            0x748f82ee,
+            0x78a5636f,
+            -0x7b3787ec,
+            -0x7338fdf8,
+            -0x6f410006,
+            -0x5baf9315,
+            -0x41065c09,
+            -0x398e870e
+    )
+
+    private val H0 =
+            intArrayOf(0x6a09e667, -0x4498517b, 0x3c6ef372, -0x5ab00ac6, 0x510e527f, -0x64fa9774, 0x1f83d9ab, 0x5be0cd19)
+
+    // Working arrays
+    private val W = IntArray(64)
+    private val H = IntArray(8)
+    private val TEMP = IntArray(8)
+
+
+    @ExperimentalUnsignedTypes
+    fun digestAsHexString(message: String): String {
+        return digest(message).toHexString()
+    }
+
+    fun digest(message: String): ByteArray {
+        return digest(message.toUtf8ByteArray())
+    }
+
+    /**
+     * Hashes the given message with SHA-256 and returns the digest.
+     *
+     * @param message The bytes to digest.
+     * @return The digest's bytes.
+     */
+    fun digest(message: ByteArray): ByteArray {
+        // Let H = H0
+        H0.arrayCopy(0, H, 0, H0.size)
+
+        // Initialize all words
+        val words = padMessage(message).toIntArray()
+
+        // Enumerate all blocks (each containing 16 words)
+        var i = 0
+        val n = words.size / 16
+        while (i < n) {
+
+            // initialize W from the block's words
+            words.arrayCopy(i * 16, W, 0, 16)
+            for (t in 16 until W.size) {
+                W[t] = (smallSig1(W[t - 2]) + W[t - 7] + smallSig0(W[t - 15]) + W[t - 16])
+            }
+
+            // Let TEMP = H
+            H.arrayCopy(0, TEMP, 0, H.size)
+
+            // Operate on TEMP
+            for (t in W.indices) {
```

Comment:
Saves creating an IntRange
```suggestion
            W.forEachIndexed { t, _ ->
```